### PR TITLE
PIMS-2044 Project List View

### PR DESCRIPTION
**Pull request by goemen**

### CODE REVIEW

File: backend/api/Areas/Project/Controllers/SearchController.cs [hunk 1]
```diff
@@ -58,7 +58,7 @@ public IActionResult GetProjectsPage()
         {
             var uri = new Uri(this.Request.GetDisplayUrl());
             var query = Microsoft.AspNetCore.WebUtilities.QueryHelpers.ParseQuery(uri.Query);
-            return GetProjectsPage(new ProjectFilterModel(query));
+            return GetProjectsPage(new ProjectFilter(query));
```
Comment (Fosol) [inline backend/api/Areas/Project/Controllers/SearchController.cs:61]: Do we need the `ProjectFilterModel` at all?  Can it be deleted if we just use the `ProjectFilter` instead?

File: backend/api/Areas/Project/Mapping/Search/ProjectPropertyMap.cs [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+                    src => src.Building != null
+                        ? GetBuildingNetValue(src.Building.Fiscals)
+                        : GetParcelNetValue(src.Parcel.Fiscals))
+                .Map(dest => dest.Estimated,
+                    src => src.Building != null
+                        ? GetBuildingEstimatedValue(src.Building.Fiscals)
+                        : GetParcelEstimatedValue(src.Parcel.Fiscals))
+
+                .Map(dest => dest.AgencyCode,
+                    src => src.Building != null ? src.Building.Agency.Parent.Code : src.Parcel.Agency.Parent.Code)
```
Comment (Fosol) [inline backend/api/Areas/Project/Mapping/Search/ProjectPropertyMap.cs:46]: Might need to check if any of the `src.Building.Agency.Parent` is null

File: backend/api/Areas/Project/Mapping/Search/ProjectPropertyMap.cs [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+                .Map(dest => dest.SubAgency,
+                    src => src.Building != null ? src.Building.Agency.Name : src.Parcel.Agency.Name)
```
Comment (Fosol) [inline backend/api/Areas/Project/Mapping/Search/ProjectPropertyMap.cs:48]: Might need to check if `src.Building.Agency` is null

File: backend/api/Areas/Project/Mapping/Search/ProjectPropertyMap.cs [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+                .Map(dest => dest.LandArea, src => src.Building != null ? src.Building.Parcel.LandArea : src.Parcel.LandArea)
```
Comment (Fosol) [inline backend/api/Areas/Project/Mapping/Search/ProjectPropertyMap.cs:50]: Might need to check if `src.Building.Parcel` is null

File: backend/api/Areas/Project/Mapping/Search/ProjectPropertyMap.cs [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+                .Inherits<BaseEntity, BaseModel>();
+
+        }
+
+        private string FormatAddress(Address address)
```
Comment (Fosol) [inline backend/api/Areas/Project/Mapping/Search/ProjectPropertyMap.cs:55]: Might want to place this in an extension method for `Address`

Comment (Fosol) [reply backend/api/Areas/Project/Mapping/Search/ProjectPropertyMap.cs:55]: All of the functions below could be useful extension methods possibly.

File: backend/api/Areas/Project/Models/Search/ProjectFilterModel.cs [hunk 6]
```diff
@@ -58,7 +64,7 @@ public ProjectFilterModel(Dictionary<string, Microsoft.Extensions.Primitives.Str
             this.Name = filter.GetStringValue(nameof(this.Name));
             this.StatusId = filter.GetIntNullValue(nameof(this.StatusId));
             this.TierLevelId = filter.GetIntNullValue(nameof(this.TierLevelId));
-
+            this.CreatedByMe = filter.GetValue<bool?>(nameof(this.CreatedByMe));
```
Comment (Fosol) [inline backend/api/Areas/Project/Models/Search/ProjectFilterModel.cs:67]: does this return `null` I can't remember if it's smart enough.  You might need to create `filter.GetBoolNullValue(...)`

File: backend/api/appsettings.Development.json [hunk 7]
```diff
@@ -9,6 +9,6 @@
     }
   },
   "ConnectionStrings": {
-    "PIMS": "Server=database,1433;User ID=admin;Database=pims"
+    "PIMS": "Server=database,5433;User ID=admin;Database=pims"
```
Comment (Fosol) [inline backend/api/appsettings.Development.json:12]: This is incorrect and should be 1433

File: backend/dal/Helpers/Extensions/ProjectExtensions.cs [hunk 8]
```diff
@@ -46,6 +85,15 @@ public static IQueryable<Entity.Project> GenerateQuery(this PimsContext context,
...
             if (filter.TierLevelId.HasValue)
                 query = query.Where(p => p.TierLevelId == filter.TierLevelId);
+            if (filter.CreatedByMe.HasValue && filter.CreatedByMe.Value)
+            {
+                query = query.Where(p => p.CreatedById.Equals(user.GetUserId()));
+            }
+
+            if (filter.Active.HasValue && filter.Active.Value)
+            {
+                query = query.Where(p => p.Status.Name != "Denied");
```
Comment (Fosol) [inline backend/dal/Helpers/Extensions/ProjectExtensions.cs:95]: We'll need to rethink this - Tech Debt

File: backend/tests/unit/dal/Core/Helpers/XmlHelperTest.cs [hunk 9]
```diff
@@ -19,7 +19,7 @@ public class XmlHelperTest
             new List<object[]>
             {
                 new object[] { new[] { new { Id = 1, Name = "test1" }, new { Id = 2, Name = "test2" } }, 2 },
-                new object[] { new[] { new Entity.TierLevel(1, "test1"), new Entity.TierLevel(2, "test2") }, 10 }
+                new object[] { new[] { new Entity.TierLevel(1, "test1"), new Entity.TierLevel(2, "test2") }, 112 }
```
Comment (Fosol) [inline backend/tests/unit/dal/Core/Helpers/XmlHelperTest.cs:22]: 112 columns!  ugh lol

File: backend/api/Areas/Project/Controllers/SearchController.cs (hunk 1 above)
Comment (goemen) [reply backend/api/Areas/Project/Controllers/SearchController.cs:61]: Not needed anymore

File: backend/api/appsettings.Development.json (hunk 7 above)
Comment (goemen) [reply backend/api/appsettings.Development.json:12]: Sorry, my local db is on PORT 5433. Will revert the change

File: backend/dal/Helpers/Extensions/ProjectExtensions.cs (hunk 8 above)
Comment (goemen) [reply backend/dal/Helpers/Extensions/ProjectExtensions.cs:95]: Are the project status order finalized? If yes, I can create extensions to check status

File: backend/api/Areas/Project/Mapping/Search/ProjectPropertyMap.cs (hunk 4 above)
Comment (devinleighsmith) [reply backend/api/Areas/Project/Mapping/Search/ProjectPropertyMap.cs:51]: This is marked as resolved but I don't see a solution here.

File: backend/api/Areas/Project/Mapping/Search/ProjectPropertyMap.cs [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+        {
+            return address != null ? $"{address.Address1} {address.Address2}".Trim() : "";
+        }
+
+        private decimal GetBuildingAssessedValue(ICollection<BuildingEvaluation> evaluations)
+        {
+            return evaluations.OrderByDescending(f => f.Date).FirstOrDefault(f => f.Key == EvaluationKeys.Assessed) ==
```
Comment (devinleighsmith) [inline backend/api/Areas/Project/Mapping/Search/ProjectPropertyMap.cs:73]: This is potentially premature optimization, but consider saving the result of the orderby so you don't have to do it twice.

File: backend/api/Areas/Project/Models/Search/ProjectModel.cs [hunk 11]
```diff
@@ -70,6 +72,21 @@ public class ProjectModel : Api.Models.BaseModel
...
+
+        /// <summary>
+        /// get/set - Project properties
+        /// </summary>
+        public IEnumerable<ProjectPropertyModel> Properties { get; set; } = new List<ProjectPropertyModel>();
+
+        /// <summary>
+        /// get/set - User Id
+        /// </summary>
+        public string UpdatedBy { get; set; }
```
Comment (devinleighsmith) [inline backend/api/Areas/Project/Models/Search/ProjectModel.cs:84]: /nit is is the User id or the display name?

File: backend/dal/Configuration/BaseEntityConfiguration`.cs [hunk 12]
```diff
@@ -17,8 +17,8 @@ public virtual void Configure(EntityTypeBuilder<TBase> builder)
             builder.Property(m => m.UpdatedOn).HasColumnType("DATETIME2");
             builder.Property(m => m.RowVersion).IsRowVersion();
 
-            builder.HasOne<User>().WithMany().HasForeignKey(m => m.CreatedById).OnDelete(DeleteBehavior.ClientSetNull);
-            builder.HasOne<User>().WithMany().HasForeignKey(m => m.UpdatedById).OnDelete(DeleteBehavior.ClientSetNull);
+            builder.HasOne(m => m.CreatedBy).WithMany().HasForeignKey(m => m.CreatedById).OnDelete(DeleteBehavior.ClientSetNull);
```
Comment (devinleighsmith) [inline backend/dal/Configuration/BaseEntityConfiguration`.cs:20]: What is the motivation for this change?

File: frontend/src/features/projects/api-service.ts [hunk 13]
```diff
@@ -0,0 +1,34 @@
+import queryString from 'query-string';
```
Comment (devinleighsmith) [inline frontend/src/features/projects/api-service.ts:1]: Not so sure about the naming of this file, any reason why it isn't just apiService.ts?

File: frontend/src/features/projects/api-service.ts [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+
+const API_ENDPOINTS = {
+  search: (filter: IProjectFilter) =>
+    `${basePath}/projects/search/page?${filter ? queryString.stringify(filter) : ''}`,
+  delete: (projectNumber: string) => `${basePath}/projects/disposal/${projectNumber}`,
+};
+
+const getProjectList = async (filter: IProjectFilter): Promise<IPagedItems<IProject>> => {
+  const url = API_ENDPOINTS.search(filter);
+  const response = await CustomAxios().get<IPagedItems<IProject>>(url);
```
Comment (devinleighsmith) [inline frontend/src/features/projects/api-service.ts:17]: This pattern works but there is no error handling here. That'll end up dumping the user back to the mapview using the existing error boundary in the event of an error, is that intended?

File: backend/dal/Configuration/BaseEntityConfiguration`.cs (hunk 12 above)
Comment (Fosol) [reply backend/dal/Configuration/BaseEntityConfiguration`.cs:20]: You should be able to update this like this;
```
builder.HasOne(m => m.CreatedBy).WithMany().HasForeignKey(m => m.CreatedById).OnDelete(DeleteBehavior.ClientSetNull);
builder.HasOne(m => m.UpdatedBy).WithMany().HasForeignKey(m => m.UpdatedById).OnDelete(DeleteBehavior.ClientSetNull);
```

File: backend/dal/Helpers/Extensions/ProjectStatusExtensions.cs [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+        {
+            return projectStatus.StatusMatch(ProjectStatusNames.Approved);
+        }
+
+        public static bool IsSubmitted(this ProjectStatus projectStatus)
+        {
+            return projectStatus.StatusMatch(ProjectStatusNames.Submitted);
+        }
+
+        public enum ProjectStatusNames
```
Comment (Fosol) [inline backend/dal/Helpers/Extensions/ProjectStatusExtensions.cs:32]: Don't introduce an enum for managing project statuses.  I choose not to lock Project Statuses to an enum so that we could avoid the pitfalls associated with this approach.  I am planning in the next sprint to introduce `ProjectStatus.Code` to use for this kind of filtering.

Comment (Fosol) [reply backend/dal/Helpers/Extensions/ProjectStatusExtensions.cs:32]: You can leave for now, but I'll have to remove next sprint.

File: backend/api/Areas/Project/Mapping/Search/ProjectPropertyMap.cs (hunk 4 above)
Comment (goemen) [reply backend/api/Areas/Project/Mapping/Search/ProjectPropertyMap.cs:51]: missed it

File: backend/api/Areas/Project/Models/Search/ProjectModel.cs (hunk 11 above)
Comment (goemen) [reply backend/api/Areas/Project/Models/Search/ProjectModel.cs:84]: DisplayName

File: frontend/src/features/projects/api-service.ts (hunk 13 above)
Comment (goemen) [reply frontend/src/features/projects/api-service.ts:1]: No reason really, I will rename it
